cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -138,74 +146,105 @@ protected boolean removeEldestEntry(Map.Entry<UUID, UserTaskInfo> eldest) {
     this(sessionExpiryMs, maxActiveUserTasks, completedUserTaskRetentionTimeMs, maxCachedCompletedUserTasks, time, new UUIDGenerator());
   }
 
+  /**
+   * Sets history limit for different endpoints. For now, only limit number of {@link EndPoint.USER_TASKS} in history to 5 entries.
+    */
+  private void initUserTaskCleanerMap() {
+    _userTaskCleanerMap.put(EndPoint.USER_TASKS, new UserTaskCleanerInfo(5));
+  }
+
+  private synchronized void cleanUserTasks() {
+    for (Map.Entry<EndPoint, UserTaskCleanerInfo> entry : _userTaskCleanerMap.entrySet()) {
+      entry.getValue().cleanUserTask();
+    }
+  }
+
   private static String httpServletRequestToString(HttpServletRequest request) {
     return String.format("%s %s", request.getMethod(), request.getRequestURI());
   }
 
   /**
-   * Create the UserTaskInfo reference if it doesn't exist.
+   * Create the UserTask reference for sync or async task if it doesn't exist.
    *
    * This method creates references {@link UserTaskInfo} and maps it to {@link HttpSession} from httpServletRequest. The
    * {@link HttpSession} is also used to fetch {@link OperationFuture} for the in-progress/completed UserTask. If the
    * UserTaskID is passed in the httpServletRequest header then that takes precedence over {@link HttpSession} to fetch
-   * the {@link OperationFuture} for the UserTask.
+   * the {@link OperationFuture} for the UserTask. Sync task is not saved in Session map.
    *
-   * @param httpServletRequest the HttpServletRequest to create the UserTaskInfo reference.
+   * @param httpServletRequest the HttpServletRequest to create the UserTaskInfo reference. Sync request should not contain
+   *                           UUID in the header.
    * @param httpServletResponse the HttpServletResponse that contains the UserTaskId in the HttpServletResponse header.
-   * @param function A function that takes a UUID and returns {@link OperationFuture}.
+   * @param function A function that takes a UUID and returns {@link OperationFuture}. For sync task, the function always
+   *                 returns a completed Future.
    * @param step The index of the step that has to be added or fetched.
+   * @param asyncComplete Indicate whether the task is sync or async.
    * @return The list of {@link OperationFuture} for the linked UserTask.
    */
-  public List<OperationFuture> getOrCreateUserTask(HttpServletRequest httpServletRequest,
-                                                   HttpServletResponse httpServletResponse,
-                                                   Function<String, OperationFuture> function,
-                                                   int step) {
+  public List<OperationFuture> getOrCreateUserTask(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+      Function<String, OperationFuture> function, int step, boolean asyncComplete) {
```

Comment:
Bad indentation -- can we ensure that we avoid making indentation errors please?